File: src/whattime_frc/MotorManagers/MotorGroup.py
from __future__ import annotations

from typing import Callable, overload
from numbers import Real

from commands2 import Command, Subsystem
from commands2 import cmd as Commands
from phoenix6.signals import NeutralModeValue

from .Motor import Motor

_UNSET = object()


class MotorGroup(Subsystem):
    def __init__(self, motors: list[Motor]) -> None:
        super().__init__()

        if len(motors) == 0:
            raise ValueError("motors cannot be empty.")

        self._motors = motors

    def toggleEnabled(self, enabled: bool | None = None) -> bool:
        allEnabled = True
        for motor in self._motors:
            allEnabled = allEnabled and motor.toggleEnabled(enabled)
        return allEnabled

    def isEnabled(self) -> bool:
        for motor in self._motors:
            if not motor.isEnabled():
                return False
        return True

    def setInverted(self, inverted: bool) -> MotorGroup:
        for motor in self._motors:
            motor.setInverted(inverted)
        return self

    def setMinValue(self, minValue: float) -> MotorGroup:
        for motor in self._motors:
            motor.setMinValue(minValue)
        return self

    def setMaxValue(self, maxValue: float) -> MotorGroup:
        for motor in self._motors:
            motor.setMaxValue(maxValue)
        return self

    def setRange(self, minValue: float, maxValue: float) -> MotorGroup:
        for motor in self._motors:
            motor.setRange(minValue, maxValue)
        return self

    def setMinSpeed(self, minSpeed: float) -> MotorGroup:
        for motor in self._motors:
            motor.setMinSpeed(minSpeed)
        return self

    def setMotorSpeed(self, motorSpeed: float) -> MotorGroup:
        for motor in self._motors:
            motor.setMotorSpeed(motorSpeed)
        return self

    def setFree(self, free: bool) -> MotorGroup:
        for motor in self._motors:
            motor.setFree(free)
        return self

    def setHoldSpeed(self, holdSpeed: float) -> MotorGroup:
        for motor in self._motors:
            motor.setHoldSpeed(holdSpeed)
        return self

    def setThreshold(self, threshold: float) -> MotorGroup:
        for motor in self._motors:
            motor.setThreshold(threshold)
        return self

    def setPG(self, pG: float) -> MotorGroup:
        for motor in self._motors:
            motor.setPG(pG)
        return self

    def setPositionSupplier(self, positionSupplier: Callable[[], float] | None) -> MotorGroup:
        for motor in self._motors:
            motor.setPositionSupplier(positionSupplier)
        return self

    @overload
    def drive(self) -> None: ...

    @overload
    def drive(self, speed: int | float) -> None: ...

    def drive(self, speed: object = _UNSET) -> None:
        if speed is _UNSET:
            for motor in self._motors:
                motor.drive()
            return

        if not isinstance(speed, Real):
            raise TypeError("speed must be an int or float.")

        scalar_speed = float(speed)
        for motor in self._motors:
            motor.drive(scalar_speed)

    def goTo(self, target: int | float) -> None:
        if not isinstance(target, Real):
            raise TypeError("target must be an int or float.")

        scalar_target = float(target)
        for motor in self._motors:
            motor.goTo(scalar_target)

    def setTarget(self, target: int | float) -> None:
        self.goTo(target)

    def goto(self, target: int | float) -> None:
        self.goTo(target)

    def setNeutralMode(self, neutralModeValue: NeutralModeValue) -> None:
        for motor in self._motors:
            motor.setNeutralMode(neutralModeValue)

    def brakelessReset(self, duration: int | float) -> Command:
        if not isinstance(duration, Real):
            raise TypeError("duration must be an int or float.")

        duration = float(duration)
        if duration < 0.0:
            raise ValueError("duration must be >= 0.")

        return Commands.sequence(
            Commands.runOnce(
                lambda: self.setNeutralMode(NeutralModeValue.COAST),
                self,
            ),
            Commands.waitSeconds(duration),
            Commands.runOnce(
                lambda: self.setNeutralMode(NeutralModeValue.BRAKE),
                self,
            ),
        )

    def getMotors(self) -> list[Motor]:
        return self._motors

File: src/main/java/com/btwrobotics/WhatTime/frc/MotorManagers/MotorGroup.java
package com.btwrobotics.WhatTime.frc.MotorManagers;

import java.util.List;
import java.util.function.DoubleSupplier;

import com.ctre.phoenix6.signals.NeutralModeValue;

import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

/**
 * Groups multiple Motor instances and applies the same actions to all of them.
 */
public class MotorGroup extends SubsystemBase {
    private final List<Motor> motors;

    public MotorGroup(List<Motor> motors) {
        if (motors == null || motors.isEmpty()) {
            throw new IllegalArgumentException("motors cannot be empty.");
        }

        this.motors = motors;
    }

    public boolean toggleEnabled() {
        return toggleEnabled(null);
    }

    public boolean toggleEnabled(Boolean enabled) {
        boolean allEnabled = true;
        for (Motor motor : motors) {
            allEnabled &= motor.toggleEnabled(enabled);
        }
        return allEnabled;
    }

    public boolean isEnabled() {
        for (Motor motor : motors) {
            if (!motor.isEnabled()) {
                return false;
            }
        }
        return true;
    }

    public MotorGroup setInverted(boolean inverted) {
        for (Motor motor : motors) {
            motor.setInverted(inverted);
        }
        return this;
    }

    public MotorGroup setMinValue(double minValue) {
        for (Motor motor : motors) {
            motor.setMinValue(minValue);
        }
        return this;
    }

    public MotorGroup setMaxValue(double maxValue) {
        for (Motor motor : motors) {
            motor.setMaxValue(maxValue);
        }
        return this;
    }

    public MotorGroup setRange(double minValue, double maxValue) {
        for (Motor motor : motors) {
            motor.setRange(minValue, maxValue);
        }
        return this;
    }

    public MotorGroup setMinSpeed(double minSpeed) {
        for (Motor motor : motors) {
            motor.setMinSpeed(minSpeed);
        }
        return this;
    }

    public MotorGroup setMotorSpeed(double motorSpeed) {
        for (Motor motor : motors) {
            motor.setMotorSpeed(motorSpeed);
        }
        return this;
    }

    public MotorGroup setFree(boolean free) {
        for (Motor motor : motors) {
            motor.setFree(free);
        }
        return this;
    }

    public MotorGroup setHoldSpeed(double holdSpeed) {
        for (Motor motor : motors) {
            motor.setHoldSpeed(holdSpeed);
        }
        return this;
    }

    public MotorGroup setThreshold(double threshold) {
        for (Motor motor : motors) {
            motor.setThreshold(threshold);
        }
        return this;
    }

    public MotorGroup setPG(double pG) {
        for (Motor motor : motors) {
            motor.setPG(pG);
        }
        return this;
    }

    public MotorGroup setPositionSupplier(DoubleSupplier positionSupplier) {
        for (Motor motor : motors) {
            motor.setPositionSupplier(positionSupplier);
        }
        return this;
    }

    public void drive() {
        for (Motor motor : motors) {
            motor.drive();
        }
    }

    public void drive(double speed) {
        for (Motor motor : motors) {
            motor.drive(speed);
        }
    }

    public void goTo(double target) {
        for (Motor motor : motors) {
            motor.goTo(target);
        }
    }

    public void setTarget(double target) {
        goTo(target);
    }

    public void setNeutralMode(NeutralModeValue neutralModeValue) {
        for (Motor motor : motors) {
            motor.setNeutralMode(neutralModeValue);
        }
    }

    public Command brakelessReset(double durationSeconds) {
        if (durationSeconds < 0.0) {
            throw new IllegalArgumentException("duration must be >= 0.");
        }

        return Commands.sequence(
            Commands.runOnce(() -> setNeutralMode(NeutralModeValue.Coast), this),
            Commands.waitSeconds(durationSeconds),
            Commands.runOnce(() -> setNeutralMode(NeutralModeValue.Brake), this)
        );
    }

    public List<Motor> getMotors() {
        return motors;
    }
}
